### PR TITLE
feat(Product Analytics): Introducing In-App Analytics Beta

### PR DESCRIPTION
**Pull request by jjoyce0510**
**Description**

This PR introduces the first version of DataHub Product Analytics, it was the result of a design sprint led by the wonderful Maggie Hays @ Spothero!

With this PR, we are happy to introduce a set of analytics that can be found under the "Analytics [beta]" tab at the top of the home page. Currently these include some _product_ analytics and some _metadata_ analytics. 

Product Analytics:

- Weekly Active Users
- Searches in the past week
- Top search terms
- Actions performed by entity type (writes) 
- Entity Details Section View views (past week) 
- Top viewed datasets (will become top viewed entities) 

Metadata Analytics:
- Number of Charts, dashboards, datasets, data tasks, data pipelines indexed by DataHub. 

**Changes**

1. Built on top of https://github.com/linkedin/datahub/pull/2435 (superseded by this PR) to send collected frontend metrics to `datahub-frontend` server
2. Adding collection endpoint to `datahub-frontend` which emits to a newly provisioned Kafka topic (DataHubUsageEvents_v1) 
3. Adding a DataHubUsageEvents consumer to listen to the product events, write into a new ElasticSearch index
4. Extended the default GMS graphQL schema in `datahub-frontend` to include 2 new Queries: getAnalyticsCharts and getHighlights. 
5. Added an AnalyticsService to `datahub-frontend` to issue analytics queries to ElasticSearch & format responses
6. Added an analytics page to the React app responsible for fetching / displaying highlights & charts

This is a first move on the analytics front, there is still much to do! Please let us know if you'd like to see any specific metrics in the dashboard :) 

**What it looks like**
Screen Shot 2021-05-04 at 11 54 32 PM
Screen Shot 2021-05-04 at 11 54 39 PM

**

**Acknowledgements** 

Big thanks to Dexter, Gabe, Shirshanka, Harshal & Maggie for the contributions! 

## Checklist
- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable)

### CODE REVIEW

File: datahub-frontend/conf/routes [hunk 1]
```diff
@@ -49,11 +49,17 @@ POST          /callback/oidc                                                  @o
...
+POST          /track                                                          react.controllers.TrackingController.track()
+
 GET           /api/*path                                                      controllers.Application.apiNotFound(path)
 POST          /api/*path                                                      controllers.Application.apiNotFound(path)
 
 # Map static resources from the /public folder to the /assets URL path
 GET           /assets/*file                                                   controllers.Assets.at(path="/public", file)
 
+# Analytics route
+POST          /track                                                          react.controllers.TrackingController.track()
```
Comment (dexter-mh-lee) [inline datahub-frontend/conf/routes:62]: Is this a duplicate of line 53??

File: datahub-web-react/.env [hunk 2]
```diff
@@ -1,2 +1,3 @@
 PUBLIC_URL=/assets
 REACT_APP_THEME_CONFIG=theme_light.config.json
+SKIP_PREFLIGHT_CHECK=true
```
Comment (dexter-mh-lee) [inline datahub-web-react/.env:3]: Is this intended?

File: datahub-frontend/play.gradle [hunk 3]
```diff
@@ -14,9 +14,9 @@ configurations {
 
 dependencies {
   if (project.hasProperty('enableEmber') && project.getProperty('enableEmber').toBoolean()) {
-    assets project(path: ':datahub-web', configuration: 'assets')
+     assets project(path: ':datahub-web', configuration: 'assets')
```
Comment (dexter-mh-lee) [inline datahub-frontend/play.gradle:17]: We should remove these extra spaces

File: datahub-frontend/build.gradle [hunk 4]
```diff
@@ -26,4 +30,20 @@ task myTar(type: Tar) {
 
 artifacts {
   archives myTar
+}
+
+graphqlCodegen {
```
Comment (dexter-mh-lee) [inline datahub-frontend/build.gradle:35]: If we can add these files via gradle task, why do we need to check the generated classes into git? Can we just gitignore them?

File: docker/datahub-frontend/Dockerfile [hunk 5]
```diff
@@ -43,5 +42,5 @@ ENV JAVA_OPTS=" \
    -Djava.security.auth.login.config=datahub-frontend/conf/jaas.conf \
    -Dlogback.configurationFile=datahub-frontend/conf/logback.xml \
    -Dlogback.debug=true \
-   -Dpidfile.path=/dev/null"
-CMD ["datahub-frontend/bin/playBinary"]
+   -Dpidfile.path=/datahub-frontend/play.pid"
```
Comment (dexter-mh-lee) [inline docker/datahub-frontend/Dockerfile:45]: Are the changes here needed?

Comment (dexter-mh-lee) [reply docker/datahub-frontend/Dockerfile:45]: This is causing smoketest to fail

File: docker/datahub-mce-consumer/env/docker.env [hunk 6]
```diff
@@ -7,4 +7,4 @@ GMS_PORT=8080
 # NOTE: Currently GMS itself does not offer SSL support, these settings are intended for when there is a proxy in front
 #       of GMS that handles SSL, such as an EC2 Load Balancer.
 #GMS_USE_SSL=true
-#GMS_SSL_PROTOCOL=
```
Comment (dexter-mh-lee) [inline docker/datahub-mce-consumer/env/docker.env:10]: Let's revert this. Not needed

File: docker/docker-compose.dev.yml [hunk 7]
```diff
@@ -42,6 +42,8 @@ services:
         APP_ENV: dev
     volumes:
       - ../datahub-frontend/build/stage/datahub-frontend:/datahub-frontend
+    ports:
+      - "9002:9002"
```
Comment (dexter-mh-lee) [inline docker/docker-compose.dev.yml:46]: Why did we add this?

Comment (jjoyce0510) [reply docker/docker-compose.dev.yml:46]: not necessarily - thanks for catching!

File: metadata-jobs/mae-consumer-job/src/main/java/com/linkedin/metadata/kafka/DataHubUsageEventsProcessor.java [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.kafka.annotation.EnableKafka;
+import org.springframework.kafka.annotation.KafkaListener;
+import org.springframework.stereotype.Component;
+
+
+@Slf4j
+@Component
+@ConditionalOnProperty(value = "DATAHUB_ANALYTICS_ENABLED", havingValue = "true", matchIfMissing = true)
```
Comment (jjoyce0510) [inline metadata-jobs/mae-consumer-job/src/main/java/com/linkedin/metadata/kafka/DataHubUsageEventsProcessor.java:22]: awesome!

File: .github/workflows/build-and-test.yml [hunk 9]
```diff
@@ -68,7 +68,9 @@ jobs:
         run: |
           ./docker/dev.sh -d
           sleep 30
-          ./smoke-test/smoke.sh
+          ./smoke-test/smoke.sh || true
+          docker ps -a
+          docker logs datahub-frontend-react
```
Comment (hsheth2) [inline .github/workflows/build-and-test.yml:73]: let's revert this now that smoke test seems to be working?

Comment (jjoyce0510) [reply .github/workflows/build-and-test.yml:73]: yes

**Pull request merged**